network/transport/v2/retry.go

```diff
@@ -0,0 +1,195 @@
+func NewPayloadRetrier(config Config, callback jobCallBack) Retriable {
+	return &payloadRetrier{
+		config:   config,
+		callback: callback,
+	}
+}
+
+type payloadRetrier struct {
+	config   Config
+	db       *bbolt.DB
+	callback jobCallBack
+	ctx      context.Context
+	cancel   context.CancelFunc
+}
+
+func (p *payloadRetrier) Configure() error {
+	p.ctx, p.cancel = context.WithCancel(context.Background())
+
+	dbFile := path.Join(p.config.Datadir, "network", "payload_jobs.db")
+	if err := os.MkdirAll(filepath.Dir(dbFile), os.ModePerm); err != nil {
+		return fmt.Errorf("unable to setup database: %w", err)
+	}
+
+	if p.config.PayloadRetryDelay == 0 {
+		p.config.PayloadRetryDelay = 5 * time.Second
+	}
+	var err error
+	p.db, err = bbolt.Open(dbFile, 0600, bbolt.DefaultOptions)
+	if err != nil {
+		return fmt.Errorf("unable to create BBolt database: %w", err)
+	}
+	if err = p.db.Update(func(tx *bbolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists([]byte("payload_jobs"))
+		return err
+	}); err != nil {
+		return fmt.Errorf("unable to create buckets in database: %w", err)
+	}
+	return nil
+}
+
+func (p *payloadRetrier) Start() error {
+	return p.db.View(func(tx *bbolt.Tx) error {
+		bucket := tx.Bucket([]byte("payload_jobs"))
+		return bucket.ForEach(func(k, v []byte) error {
+			h := hash.FromSlice(k)
+			c := binary.LittleEndian.Uint16(v)
+			p.retry(h, c)
+			return nil
+		})
+	})
+}
+
+func (p *payloadRetrier) Add(hash hash.SHA256Hash) error {
+	if err := p.writeCount(hash, 0); err != nil {
+		return err
+	}
+	p.retry(hash, 0)
+
+	return nil
+}
+
+var errContinue = errors.New("continue")
+
+func (p *payloadRetrier) retry(hash hash.SHA256Hash, initialCount uint16) {
+	startDelay := p.config.PayloadRetryDelay
+	for i := uint16(0); i < initialCount; i++ {
+		startDelay *= 2
+	}
+
+	go func() {
+		err := retry.Do(func() error {
+			count, existing, err := p.readCount(hash)
+			if err != nil {
+				return retry.Unrecoverable(err)
+			}
+			if existing {
+				if err := p.writeCount(hash, count+1); err != nil {
+					return retry.Unrecoverable(err)
+				}
+				p.callback(hash)
+				return errContinue
+			}
+
+			// no longer exists, so done
+			return nil
+		},
+			retry.Attempts(100),          // should be enough
+			retry.MaxDelay(24*time.Hour), // maximum delay of an hour
+			retry.Delay(startDelay),      // first retry after 5 seconds, second after 10, 20, 40, etc
+			retry.DelayType(retry.BackOffDelay),
+			retry.Context(p.ctx),
+			retry.LastErrorOnly(true), // only log last error
+			retry.OnRetry(func(n uint, err error) {
+				log.Logger().Debugf("retrying payload (count=%d) with ref: %s", n, hash.String())
+			}),
+		)
+		if err != nil {
+			log.Logger().Errorf("Failed to pass payload query to network: %v", err)
+		}
+	}()
+}
+
+func (p *payloadRetrier) writeCount(hash hash.SHA256Hash, count uint16) error {
+	return p.db.Update(func(tx *bbolt.Tx) error {
+		bucket := tx.Bucket([]byte("payload_jobs"))
+		data := make([]byte, 2)
```

Comment:
magic number (2), can't we get/calculate the size of a uint16 (or whatever type we're using)?

Reply:
A uint16 is 16 bits thus two bytes so it makes sense right?

Reply:
Yeah, but when somebody decides to change it and doesn't realize the `2` is related to it, things will break? (edge case though)